File: Protogen/loxim_protocol/cpp/protocol/packages/PackagesFactory.cpp
#include <stdlib.h>
#include <stdio.h>

#include "PackagesFactory.h"


#include "../base_packages/ASCPingPackage.h"
#include "../base_packages/ASCPongPackage.h"

#include "W_c_helloPackage.h"
#include "W_s_helloPackage.h"
#include "W_c_modePackage.h"
#include "W_c_loginPackage.h"
#include "W_c_authorizedPackage.h"
#include "W_c_passwordPackage.h"
#include "Q_c_statementPackage.h"
#include "Q_s_stmtparsedPackage.h"
#include "Q_c_executePackage.h"
#include "Q_s_executingPackage.h"
#include "Q_s_execution_finishedPackage.h"
#include "A_sc_okPackage.h"
#include "A_sc_errorPackage.h"
#include "A_sc_byePackage.h"
#include "A_sc_setoptPackage.h"
#include "V_sc_sendvaluesPackage.h"
#include "V_sc_sendvaluePackage.h"
#include "V_sc_finishedPackage.h"
#include "V_sc_abortPackage.h"
using namespace protocol;

Package *PackagesFactory::createPackageNotInit(uint8_t type)
{
	switch(type)
	{
	
	

		case ID_ASCPingPackage: return new ASCPingPackage();
		case ID_ASCPongPackage: return new ASCPongPackage();

		case ID_W_c_helloPackage: return new W_c_helloPackage();
		case ID_W_s_helloPackage: return new W_s_helloPackage();
		case ID_W_c_modePackage: return new W_c_modePackage();
		case ID_W_c_loginPackage: return new W_c_loginPackage();
		case ID_W_c_authorizedPackage: return new W_c_authorizedPackage();
		case ID_W_c_passwordPackage: return new W_c_passwordPackage();
		case ID_Q_c_statementPackage: return new Q_c_statementPackage();
		case ID_Q_s_stmtparsedPackage: return new Q_s_stmtparsedPackage();
		case ID_Q_c_executePackage: return new Q_c_executePackage();
		case ID_Q_s_executingPackage: return new Q_s_executingPackage();
		case ID_Q_s_execution_finishedPackage: return new Q_s_execution_finishedPackage();
		case ID_A_sc_okPackage: return new A_sc_okPackage();
		case ID_A_sc_errorPackage: return new A_sc_errorPackage();
		case ID_A_sc_byePackage: return new A_sc_byePackage();
		case ID_A_sc_setoptPackage: return new A_sc_setoptPackage();
		case ID_V_sc_sendvaluesPackage: return new V_sc_sendvaluesPackage();
		case ID_V_sc_sendvaluePackage: return new V_sc_sendvaluePackage();
		case ID_V_sc_finishedPackage: return new V_sc_finishedPackage();
		case ID_V_sc_abortPackage: return new V_sc_abortPackage();


		default:
		{
			printf("Nie znany typ pakietu");
			//TODO: Zapisz do logu informacjê o nieznanym typie pakietu 
			return NULL;
		}
	}
}

Package *PackagesFactory::createPackage(uint8_t type, char* data)
{
	Package* p=createPackageNotInit(type);
	if(!p)
		return NULL;
	if(!p->deserialize(data))
	{
		delete p;
		return NULL;
	}
	return p;
}

Package *PackagesFactory::createPackageContent(uint8_t type,PackageBufferReader* reader)
{
	Package* p=createPackageNotInit(type);
	if(!p)
		return NULL;
	if(!p->deserializeContent(reader))
	{
		delete p;
		return NULL;
	}
	return p;
};

File: Protogen/loxim_protocol/java/src/main/java/pl/edu/mimuw/loxim/protocol/packages/PackagesFactory.java
package pl.edu.mimuw.loxim.protocol.packages;

import java.util.EnumSet;
import pl.edu.mimuw.loxim.protogen.lang.java.template.ptools.Package;
import pl.edu.mimuw.loxim.protogen.lang.java.template.exception.ProtocolException;

import pl.edu.mimuw.loxim.protogen.lang.java.template.base_packages.ASCPingPackage;
import pl.edu.mimuw.loxim.protogen.lang.java.template.base_packages.ASCPongPackage;

public class PackagesFactory implements pl.edu.mimuw.loxim.protogen.lang.java.template.pstreams.PackagesFactory{

	private static PackagesFactory instance;

	public synchronized static PackagesFactory getInstance() {
		if (instance == null)
			instance = new PackagesFactory();
		return instance;
	}

	public Package createPackage(long /*short*/ package_id) throws ProtocolException {
				if (package_id==ASCPingPackage.ID) return new ASCPingPackage();
		if (package_id==ASCPongPackage.ID) return new ASCPongPackage();
										if (package_id==W_c_helloPackage.ID) return new W_c_helloPackage();
											if (package_id==W_s_helloPackage.ID) return new W_s_helloPackage();
											if (package_id==W_c_modePackage.ID) return new W_c_modePackage();
											if (package_id==W_c_loginPackage.ID) return new W_c_loginPackage();
											if (package_id==W_c_authorizedPackage.ID) return new W_c_authorizedPackage();
											if (package_id==W_c_passwordPackage.ID) return new W_c_passwordPackage();
											if (package_id==Q_c_statementPackage.ID) return new Q_c_statementPackage();
											if (package_id==Q_s_stmtparsedPackage.ID) return new Q_s_stmtparsedPackage();
											if (package_id==Q_c_executePackage.ID) return new Q_c_executePackage();
											if (package_id==Q_s_executingPackage.ID) return new Q_s_executingPackage();
											if (package_id==Q_s_execution_finishedPackage.ID) return new Q_s_execution_finishedPackage();
											if (package_id==A_sc_okPackage.ID) return new A_sc_okPackage();
											if (package_id==A_sc_errorPackage.ID) return new A_sc_errorPackage();
											if (package_id==A_sc_byePackage.ID) return new A_sc_byePackage();
											if (package_id==A_sc_setoptPackage.ID) return new A_sc_setoptPackage();
											if (package_id==V_sc_sendvaluesPackage.ID) return new V_sc_sendvaluesPackage();
											if (package_id==V_sc_sendvaluePackage.ID) return new V_sc_sendvaluePackage();
											if (package_id==V_sc_finishedPackage.ID) return new V_sc_finishedPackage();
											if (package_id==V_sc_abortPackage.ID) return new V_sc_abortPackage();
							throw new ProtocolException("Cannot create instance of package with id:"+package_id+". Unknown package.");
	}
}


